src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -367,24 +392,128 @@ private static String getClassNameFromQualifiedName(String qualifiedName) {
+     * considered to be super class of an anonymous inner class.
+     *
+     * <p>
+     * Suppose our pattern class is {@code Main.ClassOne} and class to be matched is
+     * {@code Main.ClassOne.ClassTwo.ClassThree} then type declaration name matching count would
+     * be calculated by comparing every character, and updating main counter when we hit "." or
+     * when it is the last character of the pattern class and certain conditions are met. This is
+     * done so that matching count is 13 instead of 5. This is due to the fact that pattern class
+     * can contain anonymous inner class object of a nested class which isn't true in case of
+     * extending classes as you can't extend nested classes.
+     * </p>
+     *
+     * @param patternTypeDeclaration type declaration against which the given type declaration has
+     *                               to be matched
+     * @param typeDeclarationToBeMatched type declaration to be matched
+     * @return type declaration matching count
+     */
+    private static int getAnonSuperTypeMatchingCount(String patternTypeDeclaration,
+                                                    String typeDeclarationToBeMatched) {
+        final int minLength = Math
+            .min(typeDeclarationToBeMatched.length(), patternTypeDeclaration.length());
+        final char packageSeparator = PACKAGE_SEPARATOR.charAt(0);
+        int result = 0;
+        final boolean shouldCountBeUpdatedAtLastCharacter =
+            typeDeclarationToBeMatched.length() > minLength
+                && typeDeclarationToBeMatched.charAt(minLength) == packageSeparator;
+
+        for (int index = 0; index < minLength
+            && patternTypeDeclaration.charAt(index) == typeDeclarationToBeMatched.charAt(index);
+             ++index) {
+            if (index == minLength - 1
+                && shouldCountBeUpdatedAtLastCharacter
+                || patternTypeDeclaration.charAt(index) == packageSeparator) {
+                result = index;
+            }
+        }
+        return result;
+    }
 
-        /** Corresponding node. */
-        private final DetailAST classAst;
+    /**
+     * Maintains information about the type of declaration.
+     * Any ast node of type {@link TokenTypes#CLASS_DEF} or {@link TokenTypes#INTERFACE_DEF}
+     * or {@link TokenTypes#ENUM_DEF} or {@link TokenTypes#ANNOTATION_DEF}
+     * or {@link TokenTypes#RECORD_DEF} is considered as a type declaration.
+     */
+    private static class TypeDeclarationDescription {
 
-        /** Qualified class name(with package). */
+        /**
+         * Complete type declaration name with package name and outer type declaration name.
+         */
         private final String qualifiedName;
 
+        /**
+         * Depth of nesting of type declaration. The depth of top-level class is 0,
+         * a class nested under it has a depth of -1 and so on.
+         */
+        private final int depth;
```

Comment:
Why is depth negative?

Reply:
Depth is always compared in an inverted manner, this change was done to use method ref-
```
.max(Comparator
    .comparingInt(countProvider::apply)
    .thenComparingInt(ClassDesc::getDepth)) // <--- Over here
```
It looks cleaner compared to using lambda and explicitly using a minus sign.

Reply:
> looks cleaner

I am not sure this is a good reason to do something, if it is not logical :) Wouldn't it make sense it have depth be positive, and use `.min` since we are finding the nearest class?

Reply:
@Vyom-Yadav ping

Reply:
> I am not sure this is a good reason to do something, if it is not logical :) Wouldn't it make sense it have depth be positive, and use .min since we are finding the nearest class?

Using `.max` shouldn't be a problem as we are finding the class with the most matching count, having negative depth shouldn't be a problem, the deeper you go the more negative it becomes.

Reply:
> the deeper you go the more negative it becomes.

This is exactly my point; this is not intuitive, and you could just use `min` and reverse the sorting order, this still allows fancy syntax, method references, etc. One does not say "we are negative five feet deep in the water" similar to how it would not make sense to say that "we are negative three nested classes deep" .

Think about next maintainer that comes along; when we use depth in the real world, if it is in negative terms, this would be "elevation". Do other checks that use `depth` consider depth to be negative? It does not appear that depth was negative in this check previously either.

Reply:
It won't be possible to use `Comparator.comparingInt(..)` with method references because either depth or matching count needs to be with the opposite sign. Inverting the sign of matching count isn't intuitive as well, so I reverted back to our original approach.

Reply:
Resolved at https://github.com/checkstyle/checkstyle/pull/11468#discussion_r873170995